File: RpnCalculator-Scala/src/rpn/scala/operators/FactorialOperator.scala
package rpn.scala.operators

class FactorialOperator extends UnaryOperator {
  val FACTORIAL = "!"

  def doOperation(value: Int): Int = {
    factorial(value)
  }

  def factorial(value: Int): Int = {
    if (value == 0) 1
    else value * doOperation(value - 1)
  }

  def handlesOperatorCharacter(operand: String): Boolean = { operand == FACTORIAL }

}

File: src/main/java/rpn/operator/FactorialOperator.java
package rpn.operator;

import rpn.RpnStack;

public class FactorialOperator implements Operator {

    private static final String FACTORIAL = "!";

    @Override
    public int doOperation(RpnStack numbers) {
        int total = 1;
        Integer operand = numbers.pop();
        while (operand > 0) {
            total = total * operand;
            operand--;
        }
        numbers.push( total );
        return total;
    }

    @Override
    public boolean handlesOperatorCharacter(String operand) {
        return FACTORIAL.contentEquals( operand );
    }

}
